File: palindrome.dart
import 'dart:io';

var palindrome = "";
var freq = new Map();

Map countLetters(String string) {
  for (var letter in string.split('')) {
    if (freq.containsKey(letter)) {
      ++freq[letter];
    } else {
      freq[letter] = 1;
    }
  }
  return freq;
}

Bool isPalindrome(String string) {
  return (string == string.split('').reversed.join(''))
    ? true
    : false;
}

Bool canBePalindrome(Map freq) {
  var numberOfOdds = 0;
  for (var amount in freq.values) {
    if (amount % 2 != 0) {
      ++numberOfOdds;
    }
  }
  return (numberOfOdds <= 1)
    ? true
    : false;
}

void findOdd(Symbol k, Num v) {
  if (v % 2 != 0) {
    palindrome = palindrome + k.toString();
    freq[k] -= 1;
  }
}

String combineParts(Symbol k, Num v) {
  palindrome = palindrome + (k * (v / 2).floor());
  palindrome = (k * (v / 2).floor()) + palindrome;
}

String createPalindrome(Map freq) {
  freq.forEach((k, v) => findOdd(k, v));
  freq.forEach((k, v) => combineParts(k, v));
  return palindrome;
}

void main() {
  stdout.writeln("Enter a word or phrase");
  String input = stdin.readLineSync();
  freq = countLetters(input);
  if (isPalindrome(input)) {
      print("$input is already a palindrome.");
  } else if (!canBePalindrome(freq)) {
      print("$input cannot be made into a palindrome.");
  } else {
      print("Palindrome of $input: ${createPalindrome(freq)}");
  }
}

File: palindrome.py
def count_letters(string):
    freq, limit, i = {}, len(string), 0
    while i < limit:
        if string[i] in freq:
            freq[string[i]] += 1
        else:
            freq[string[i]] = 1
        i += 1
    return freq

def is_palindrome(string):
    return True if string == string[::-1] else False

def can_be_palindrome(freq):
    numberOfOdds = 0
    for num in freq.values():
        if num % 2 != 0:
            numberOfOdds += 1
    if numberOfOdds > 1:
        return False
    else:
        return True

def create_palindrome(freq):
    palindrome = ""
    for (let, num) in freq.items():
        if num % 2 != 0:
            palindrome += let
            freq[let] -= 1
    for (let, num) in freq.items():
        palindrome = (let * (int(round(num / 2)))) + palindrome + (let * (int(round(num / 2))))
    return palindrome

def main(string):
    freq = count_letters(string)
    if is_palindrome(string):
        print ("{} is already a palindrome.".format(string))
    elif not can_be_palindrome(freq):
        print ("{} cannot be made into a palindrome.".format(string))
    else:
        palindrome = create_palindrome(freq)
        print ("Palindrome of {}: {}".format(string, palindrome))
